feat: distinguish failure to start during upgrade

jacderida opened the pull request:
It's possible for a service to be upgraded but then not subsequently start. In this case, it has still been upgraded to a new version. It's worth making a distinction to the user between an actual error in the upgrade process, or a failure to start, because in the latter case, they do actually have an upgrade. They can then take action to try and start their services again.

As part of this change, the start process attempts to find whether the service process did indeed start, because you don't always seem to get errors back from the service infrastructure. We also make sure that we return an error if there was a failure with the upgrade process for any services. This is necessary for visibility on our own deploy process.

## Description

reviewpad:summary

joshuef commented on sn_node/src/node/flow_ctrl/mod.rs:
```diff
@@ -69,16 +70,22 @@ impl FlowCtrl {
     }
 
     /// Process the next pending cmds
```
Lets add a bit of documentation around the semaphore and limit, and it's intention?

joshuef commented on sn_node/src/node/flow_ctrl/mod.rs:
```diff
@@ -139,6 +146,8 @@ impl FlowCtrl {
         let mut last_link_cleanup = Instant::now();
         let mut last_dysfunction_check = Instant::now();
 
+        let semaphore = Arc::new(Semaphore::new(5));
```
lets make this value a `const` and get some docs around it/why it was chosen. (eventually we can make it configurable, but for now I think it's fine)

joshuef commented:
Closing as this is likely not our solution

jacderida committed:
feat: distinguish failure to start during upgrade

It's possible for a service to be upgraded but then not subsequently start. In this case, it has
still been upgraded to a new version. It's worth making a distinction to the user between an actual
error in the upgrade process, or a failure to start, because in the latter case, they do actually
have an upgrade. They can then take action to try and start their services again.

As part of this change, the start process attempts to find whether the service process did indeed
start, because you don't always seem to get errors back from the service infrastructure. We also
make sure that we return an error if there was a failure with the upgrade process for any services.
This is necessary for visibility on our own deploy process.
